copy/copy.go

```diff
@@ -699,24 +699,61 @@ func checkImageDestinationForCurrentRuntime(ctx context.Context, sys *types.Syst
 		if err != nil {
 			return errors.Wrapf(err, "Error parsing image configuration")
 		}
-
-		wantedOS := runtime.GOOS
-		if sys != nil && sys.OSChoice != "" {
-			wantedOS = sys.OSChoice
+		platforms, err := platform.WantedPlatforms(sys)
+		if err != nil {
+			return errors.Wrapf(err, "error getting current platform information %#v", sys)
 		}
-		if wantedOS != c.OS {
-			logrus.Infof("Image operating system mismatch: image uses %q, expecting %q", c.OS, wantedOS)
+		warning := runtimeMatchWarning(c, platforms)
+		if warning != "" {
+			logrus.Info(warning)
 		}
+	}
+	return nil
+}
 
-		wantedArch := runtime.GOARCH
-		if sys != nil && sys.ArchitectureChoice != "" {
-			wantedArch = sys.ArchitectureChoice
+// runtimeMatchWarning returns an error message warning about the mismatch
+// between an image config and set of platforms by a runtime, or "" if the two match.
+func runtimeMatchWarning(cfg *imgspecv1.Image, platforms []imgspecv1.Platform) string {
+	// We want to report a mismatch only if _none_ of platforms matches.
+	// In the most general case, we can only list all of the available platforms; let’s try
+	// to do better in the common cases.
+	osOptions := newOrderedSet()
+	archOptions := newOrderedSet()
+	var matches bool
+	for _, wantedPlatform := range platforms {
+		// Waiting for https://github.com/opencontainers/image-spec/pull/777 :
+		// This currently can’t use image.MatchesPlatform because we don’t know what to use
+		// for image.Variant.
+		matches = true
+		if wantedPlatform.OS != cfg.OS {
+			osOptions.append(wantedPlatform.OS)
+			matches = false
+		}
+		if wantedPlatform.Architecture != cfg.Architecture {
+			archOptions.append(wantedPlatform.Architecture)
+			matches = false
+		}
+		if matches {
+			break
 		}
-		if wantedArch != c.Architecture {
-			logrus.Infof("Image architecture mismatch: image uses %q, expecting %q", c.Architecture, wantedArch)
+	}
+	switch {
+	case matches:
+		return "" // OK
+	case len(osOptions.list) != 0 && len(archOptions.list) == 0:
+		return fmt.Sprintf("Image operating system mismatch: image uses %q, expecting one of %q",
+			cfg.OS, strings.Join(osOptions.list, ", "))
+	case len(osOptions.list) == 0 && len(archOptions.list) != 0:
+		return fmt.Sprintf("Image architecture mismatch: image uses %q, expecting one of %q",
+			cfg.Architecture, strings.Join(archOptions.list, ", "))
+	default:
+		options := newOrderedSet()
+		for _, o := range platforms {
+			options.append(fmt.Sprintf("%s+%s", o.OS, o.Architecture))
 		}
+		return fmt.Sprintf("Image operating system mismatch: image uses OS %q+architecture %q, expecting one of %q",
```

Comment:
I prefer to only keep the `default` case here. I don't like the precedence of OS over the arch as both could very well mismatch.

Reply:
It would simply this function quite a bit as well.

Reply:
Sure, the default can be good enough.

In that case I’m tempted to inline the warning generator back and drop the tests. WDYT?

Reply:
SGTM :+1: